go/adbc/driver/snowflake/connection.go

```diff
@@ -557,13 +587,204 @@ func (c *cnxn) populateMetadata(ctx context.Context, depth adbc.ObjectDepth, cat
+		}
+
+		currentConstraintQualifiedName := getFullyQualifiedConstraintName(uniqueConstraint.dbName, uniqueConstraint.schema, uniqueConstraint.tblName, uniqueConstraint.constraintName)
+
+		// skip constraint if it doesn't exist in fullyQualifiedConstraints
+		if _, exists := fullyQualifiedConstraints[currentConstraintQualifiedName]; exists {
+			uniqueConstraint.constraintType = internal.Unique
+			uniqueConstraintsData = append(uniqueConstraintsData, uniqueConstraint)
+		}
+
+	}
+	return uniqueConstraintsData, nil
+}
+
+func (c *cnxn) getPrimaryKeyConstraints(ctx context.Context, fullyQualifiedConstraints map[string]bool) ([]TableConstraint, error) {
+	primaryKeyConstraintsData := make([]TableConstraint, 0)
+
+	rows, err := c.sqldb.QueryContext(ctx, preparePrimaryKeyConstraintSQL(), nil)
+	if err != nil {
+		return nil, errToAdbcErr(adbc.StatusIO, err)
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var primaryKeyConstraint TableConstraint
+		if err := rows.Scan(&primaryKeyConstraint.skipCreatedOn, &primaryKeyConstraint.dbName, &primaryKeyConstraint.schema, &primaryKeyConstraint.tblName, &primaryKeyConstraint.colName, &primaryKeyConstraint.keySequence, &primaryKeyConstraint.constraintName, &primaryKeyConstraint.skipRely, &primaryKeyConstraint.skipComment); err != nil {
+			return nil, errToAdbcErr(adbc.StatusInvalidData, err)
+		}
+
+		currentConstraintQualifiedName := getFullyQualifiedConstraintName(primaryKeyConstraint.dbName, primaryKeyConstraint.schema, primaryKeyConstraint.tblName, primaryKeyConstraint.constraintName)
+
+		// skip constraint if it doesn't exist in fullyQualifiedConstraints
+		if _, exists := fullyQualifiedConstraints[currentConstraintQualifiedName]; exists {
+			primaryKeyConstraint.constraintType = internal.PrimaryKey
+			primaryKeyConstraintsData = append(primaryKeyConstraintsData, primaryKeyConstraint)
+		}
+
+	}
+	return primaryKeyConstraintsData, nil
+}
+
+func (c *cnxn) getForeignKeyConstraints(ctx context.Context, fullyQualifiedConstraints map[string]bool) ([]TableConstraint, error) {
+	foreignKeyConstraintsData := make([]TableConstraint, 0)
+
+	rows, err := c.sqldb.QueryContext(ctx, prepareForeignKeyConstraintSQL(), nil)
+	if err != nil {
+		return nil, errToAdbcErr(adbc.StatusIO, err)
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var foreignKeyConstraint TableConstraint
+		if err := rows.Scan(&foreignKeyConstraint.skipCreatedOn, &foreignKeyConstraint.dbName, &foreignKeyConstraint.schema, &foreignKeyConstraint.tblName, &foreignKeyConstraint.colName, &foreignKeyConstraint.fkDbName, &foreignKeyConstraint.fkSchema, &foreignKeyConstraint.fkTblName, &foreignKeyConstraint.fkColName, &foreignKeyConstraint.keySequence, &foreignKeyConstraint.skipUpdateRule, &foreignKeyConstraint.skipDeleteRule, &foreignKeyConstraint.fk_constraintName, &foreignKeyConstraint.constraintName, &foreignKeyConstraint.skipDeferrability, &foreignKeyConstraint.skipRely, &foreignKeyConstraint.skipComment); err != nil {
```

Comment:
there's *gotta* be a more readable way to do this....

Reply:
split into multiple lines